### PR TITLE
Feat/odw 1420 curatedlayer nsip s51 advice

### PR DESCRIPTION
**Pull request by KranthiRayipudi**
**PR Template**

 1. Are there new source to raw datasets?
	
	 - [ ] If yes - has a trigger been attached at the appropriate frequency?
  	 - [ ] No

 2. Have any new tables been created in Standardised?

  	- [ ] No
   	- [ ] If yes:
		- [ ] orchestration.json has been updated and tested in Dev and has been / is about to be PRd into main
		- [ ] the new schema exists inside *odw-config/standardised-table-definitions* OR is about to be PRd into main
			- Make sure to run Platform Integrate and Platform Deploy to Dev at least to ensure the schema is deployed into Synapse Dev Live 
		- [ ] Is the raw-to-standardised python script scheduled to run for this dataset grouping?
 3. Have any new tables been created in Harmonised or Curated?

   	 - [ ] No
     	 - [ ] If yes
	 	- [ ]  *2-odw-standardised-to-harmonised/py_odw_harmonised_table_creation* OR 4-*odw-harmonised-to-curated/py_odw_curated_table_creation* are set up to run in the pipeline *pln_post_deployments* with the base parameter specifying the correct table
		- [ ] the new schema exists inside *odw-config/harmonised-table-definitions* / *odw-config/curated-table-definitions* OR is about to be PRd into main
			- Make sure to run Platform Integrate and Platform Deploy to Dev at least to ensure the schema is deployed into Synapse Dev Live 
 4. Have any tables changed AND/OR have any columns changed in any scripts?
We only care about new columns or columns that change type.
	- [ ] No
 	- [ ] If yes:
		- [ ] Please set py_change_table to run in the pipeline *pln_post_deployments*
		- [ ] Please create a script to backdate and fill in this new column in Test and Prod
			Only delete and recreate tables with caution!
 4. Have any scripts run in isolation in dev? Please look at the *"spark.autotune.trackingId"*
		- If these changes need to be reflected in Test and Prod, please add to the pipeline *post-			deployment/pln_post_deployment*
	- [ ] Yes - I have reflected this script in the *pln_post_deployments* pipeline
	- [ ] Yes - This script is part of a scheduled run and has been added to the appropriate end to end pipeline with a trigger at the correct frequency
	- [ ] No - This change does not need to take place in Test and Prod
	- [ ] No - No scripts have run

### CODE REVIEW

File: workspace/notebook/py_unit_tests_functions.json [hunk 1]
```diff
@@ -469,7 +469,7 @@
 					}
 				},
 				"source": [
```
Comment (muhammadkc) [inline workspace/notebook/py_unit_tests_functions.json:471]: The function test_hrm_to_curated_no_dropping_primary_ids_only_records is being used in all unit test notebooks and removing the default value might break other notebooks. Can I please ask why it's been removed and if can be put back?

File: workspace/notebook/py_unit_tests_functions.json [hunk 2]
```diff
@@ -881,6 +881,44 @@
...
+					"nteract": {
+						"transient": {
+							"deleting": false
+						}
+					}
+				},
+				"source": [
+					"from collections import Counter\r\n",
+					"\r\n",
+					"def Array_duplicates(curated_db: str, curated_table: str, sourceColumn: str):\r\n",
```
Comment (muhammadkc) [inline workspace/notebook/py_unit_tests_functions.json:901]: Can we follow the naming convention for naming test functions please? A more relevant name could be `test_duplicates_in_array_column`

File: workspace/notebook/py_unit_tests_nsip_s51_advice.json [hunk 3]
```diff
@@ -855,7 +855,72 @@
...
+					},
+					"nteract": {
+						"transient": {
+							"deleting": false
+						}
+					}
+				},
+				"source": [
+					"result = Array_duplicates(curated_db_name, curated_table_name,\"attachmentIds\")\r\n",
+					"print(f\"duplicates found in the attachmentIds: {result}\")"
```
Comment (muhammadkc) [inline workspace/notebook/py_unit_tests_nsip_s51_advice.json:889]: increment exitCode if the test fails

File: workspace/notebook/py_unit_tests_nsip_s51_advice.json [hunk 4]
```diff
@@ -855,7 +855,72 @@
...
+					},
+					"nteract": {
+						"transient": {
+							"deleting": false
+						}
+					}
+				},
+				"source": [
+					"result = Array_duplicates(curated_db_migration_name, curated_table_migration_name,\"attachmentIds\")\r\n",
+					"print(f\"duplicates found in the attachmentIds: {result}\")"
```
Comment (muhammadkc) [inline workspace/notebook/py_unit_tests_nsip_s51_advice.json:921]: increment exitCode if the test fails

File: workspace/notebook/py_unit_tests_functions.json (hunk 1 above)
Comment (KranthiRayipudi) [reply workspace/notebook/py_unit_tests_functions.json:471]: I have added default value test_hrm_to_curated_no_dropping_primary_ids_only_records  function

File: workspace/notebook/py_unit_tests_functions.json (hunk 2 above)
Comment (KranthiRayipudi) [reply workspace/notebook/py_unit_tests_functions.json]: I renamed function name

File: workspace/notebook/py_unit_tests_functions.json [hunk 5]
```diff
@@ -881,6 +881,44 @@
...
+					"                                      {curated_db}.{curated_table}\"\"\").collect()  \r\n",
+					"   # Assuming there's only one Row object in the list\r\n",
+					"    row = attachmentIds_list[0]    \r\n",
+					"    # Get the attachmentIds\r\n",
+					"    attachment_ids = row.ids    \r\n",
+					"    # Count occurrences of each attachmentId\r\n",
+					"    id_counts = Counter(attachment_ids)    \r\n",
+					"    # Print duplicates\r\n",
+					"    duplicates_found = False\r\n",
+					"    for id, count in id_counts.items():\r\n",
```
[the remark this answers is not in the document]
Comment (KranthiRayipudi) [reply workspace/notebook/py_unit_tests_functions.json:912]: I changed code

File: workspace/notebook/py_unit_tests_nsip_s51_advice.json (hunk 3 above)
Comment (KranthiRayipudi) [reply workspace/notebook/py_unit_tests_nsip_s51_advice.json]: I have added exitCode

File: workspace/notebook/py_unit_tests_nsip_s51_advice.json (hunk 4 above)
Comment (KranthiRayipudi) [reply workspace/notebook/py_unit_tests_nsip_s51_advice.json]: I have added exitCode

**Review by muhammadkc — APPROVED** [review]

**Pull request merged**